### PR TITLE
Utilize array length index for `= []`

### PR DESCRIPTION
**Pull request by jeeminso**
## Summary of the changes / Why this improves CrateDB
For arrays from tables created on or after `5.9` can utilize `_array_length_` index. For `= []`, `_array_length_ = 0` should be enough, but for non empty arrays(i.e. `= [1]`), array length queries [did not improve the performance.](https://github.com/crate/crate/pull/16479#issuecomment-2310825868)

## Checklist

 - [x] Added an entry in the latest `docs/appendices/release-notes/<x.y.0>.rst` for user facing changes
 - [x] Updated documentation & `sql_features` table for user facing changes
 - [x] Touched code is covered by tests
 - [x] [CLA](https://crate.io/community/contribute/cla/) is signed
 - [x] This does not contain breaking changes, or if it does:
    - It is released within a major release
    - It is recorded in the latest `docs/appendices/release-notes/<x.y.0>.rst`
    - It was marked as deprecated in an earlier release if possible
    - You've thought about the consequences and other components are adapted
      (E.g. AdminUI)

### CODE REVIEW

**Review by mfussenegger — COMMENTED** [review]
Could you also run some benchmarks?

I'm not sure if there'll be much/any benefit for the cases where we already use a  TermInSetQuery or PointInSetQuery

If there's a big improvement it could be mentioned in the release notes

File: server/src/main/java/io/crate/expression/operator/EqOperator.java [hunk 1]
```diff
@@ -193,34 +195,55 @@ private static Query termsAndGenericFilter(Function function,
...
-        if (values.isEmpty()) {
-            // `arrayRef = []` - termsQuery would be null
+        if (flattenUniqueValues.isEmpty()) { // using flattenUnique to catch nested empty arrays
+            if (canUseArrayLengthIndex) {
+                return new IntEqQuery().termQuery(toArrayLengthFieldName(
+                        Objects.requireNonNull(context.tableInfo().getReference(column)),
+                        context.tableInfo()::getReference),
+                    0,
+                    true,
+                    true);
```
[the remark this answers is not in the document]
Comment (mfussenegger) [reply server/src/main/java/io/crate/expression/operator/EqOperator.java:209]: Or another variant would be to hide the `toArrayLengthFieldName`, replacing it with a `arrayLengthQuery` method on the ArrayIndexer to keep both write/read details encapsulated in the same class

**Comment by jeeminso** [comment]
Thanks for pointing out, I assumed that utilizing the array length index was always better for performance. I ran the benchmarks are the results are as follows:

```
# Results (server side duration in ms)
V1: 5.9.0-58b054dc47159ca389384d1a20ada3bc94efc8ee
V2: 5.9.0-4124f4da5428009196ff09257dbb3c0784394ad7

Q: select count(*) from t_arrays_biased_empty where xs = []
C: 1
| Version |         Mean ±    Stdev |        Min |     Median |         Q3 |        Max |
|   V1    |       85.133 ±  363.339 |     44.963 |     46.384 |     47.271 |   5019.434 |
|   V2    |        8.697 ±   71.904 |      2.484 |      3.367 |      3.843 |   1020.350 |
├---------┴-------------------------┴------------┴------------┴------------┴------------┘
|               - 162.92%                           - 172.93%   
There is a 99.63% probability that the observed difference is not random, and the best estimate of that difference is 162.92%
The test has statistical significance

Q: select count(*) from t_arrays_biased_5_elements where xs = []
C: 1
| Version |         Mean ±    Stdev |        Min |     Median |         Q3 |        Max |
|   V1    |       34.612 ±  233.043 |     13.774 |     14.379 |     14.766 |   3293.747 |
|   V2    |        2.552 ±    2.528 |      1.776 |      2.272 |      2.466 |     37.490 |
├---------┴-------------------------┴------------┴------------┴------------┴------------┘
|               - 172.53%                           - 145.42%   
There is a 94.76% probability that the observed difference is not random, and the best estimate of that difference is 172.53%
The test has no statistical significance

Q: select count(*) from t_arrays_biased_empty where xs = [1,1]
C: 1
| Version |         Mean ±    Stdev |        Min |     Median |         Q3 |        Max |
|   V1    |        1.406 ±    1.344 |      0.959 |      1.256 |      1.401 |     19.689 |
|   V2    |        2.991 ±    1.633 |      2.084 |      2.787 |      3.034 |     24.816 |
├---------┴-------------------------┴------------┴------------┴------------┴------------┘
|               +  72.08%                           +  75.77%   
There is a 100.00% probability that the observed difference is not random, and the best estimate of that difference is 72.08%
The test has statistical significance

Q: select count(*) from t_arrays_biased_5_elements where xs = [1,1]
C: 1
| Version |         Mean ±    Stdev |        Min |     Median |         Q3 |        Max |
|   V1    |        1.081 ±    0.219 |      0.793 |      1.029 |      1.176 |      2.113 |
|   V2    |        2.299 ±    0.269 |      1.821 |      2.259 |      2.395 |      3.706 |
├---------┴-------------------------┴------------┴------------┴------------┴------------┘
|               +  72.03%                           +  74.85%   
There is a 100.00% probability that the observed difference is not random, and the best estimate of that difference is 72.03%
The test has statistical significance

Q: select count(*) from t_arrays_biased_empty where xs = [1,1,1,1,1]
C: 1
| Version |         Mean ±    Stdev |        Min |     Median |         Q3 |        Max |
|   V1    |        1.120 ±    0.280 |      0.783 |      1.062 |      1.194 |      3.393 |
|   V2    |        2.966 ±    1.381 |      2.327 |      2.813 |      2.947 |     20.987 |
├---------┴-------------------------┴------------┴------------┴------------┴------------┘
|               +  90.38%                           +  90.36%   
There is a 100.00% probability that the observed difference is not random, and the best estimate of that difference is 90.38%
The test has statistical significance

Q: select count(*) from t_arrays_biased_5_elements where xs = [1,1,1,1,1]
C: 1
| Version |         Mean ±    Stdev |        Min |     Median |         Q3 |        Max |
|   V1    |        0.901 ±    0.191 |      0.664 |      0.842 |      0.948 |      1.676 |
|   V2    |        2.546 ±    0.375 |      2.119 |      2.492 |      2.656 |      5.966 |
├---------┴-------------------------┴------------┴------------┴------------┴------------┘
|               +  95.45%                           +  99.01%   
There is a 100.00% probability that the observed difference is not random, and the best estimate of that difference is 95.45%
The test has statistical significance
```

Notice that `select count(*) from t_arrays_biased_5_elements where xs = [1,1]`, the array length index must pre-filter all rows since non of them should be length=2 but the performance actually degraded. I have done a few followup experiments but could not observe any improvements.

I will keep the changes for `=[]` but revert non empty array case.

**Comment by jeeminso** [comment]
retest this please
unrelated failure, https://jenkins.crate.io/job/CrateDB/job/crate_on_pr/16423/testReport/junit/io.crate.integrationtests/TransportSQLActionTest/test_types_with_storage_can_be_inserted_and_queried/

File: server/src/main/java/io/crate/expression/operator/EqOperator.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (jeeminso) [reply server/src/main/java/io/crate/expression/operator/EqOperator.java]: Thanks, added `arrayLengthQuery` methods.

**Comment by mfussenegger** [comment]
> Thanks for pointing out, I assumed that utilizing the array length index was always better for performance. I ran the benchmarks and the results are as follows

You could also try replacing the ` NumTermsPerDocQuery` with the array-length query instead of adding both for the pre-filtering

**Comment by jeeminso** [comment]
> You could also try replacing the  `NumTermsPerDocQuery` with the array-length query instead of adding both for the pre-filtering

`NumTermsPerDocQuery` is only used for arrays before `5.9` where array length indexes are not available, could you double check?

**Comment by jeeminso** [comment]
> You could also try replacing the  NumTermsPerDocQuery with the array-length query instead of adding both for the pre-filtering

`NumTermsPerDocQuery` is only used for arrays before `5.9` when array length indexes are not available, could you please double check?

**Review by mfussenegger — APPROVED** [review]

**Pull request merged**